File: ios/Sources/NetworkToolsPlugin/SessionRegistry.swift
import Foundation
import Security

/**
 * Manages stateful session lifecycles with vaulted credential storage and TTL-based eviction.
 *
 * Credentials are stored in the iOS Keychain (kSecAttrAccessibleWhenUnlocked).
 * An in-memory index tracks last-used timestamps for TTL enforcement without loading
 * sensitive data into plain Swift structures unnecessarily.
 *
 * Generic by design — reusable for SSH, TCP, WebSocket, or any stateful connection.
 */
class SessionRegistry {
    private let namespace: String
    private let ttlSeconds: TimeInterval
    private var lastUsed: [String: Date] = [:]
    private let lock = NSLock()

    private let keychainService = "com.t6x.networktools.session_registry"

    init(namespace: String, ttlSeconds: TimeInterval = 900) {
        self.namespace = namespace
        self.ttlSeconds = ttlSeconds
    }

    // MARK: - Public API

    /** Vault the connection params for sessionId and record its last-used time. */
    func register(sessionId: String, params: [String: Any?]) {
        evictExpired()
        let filtered = params.compactMapValues { $0 }
        guard let data = try? JSONSerialization.data(withJSONObject: filtered) else { return }
        saveToKeychain(key: keychainKey(sessionId), data: data)
        lock.lock()
        lastUsed[sessionId] = Date()
        lock.unlock()
    }

    /**
     * Retrieve vaulted params for sessionId.
     * Returns nil if the session is unknown or has exceeded the TTL.
     */
    func getParams(sessionId: String) -> [String: Any]? {
        evictExpired()
        lock.lock()
        let last = lastUsed[sessionId]
        lock.unlock()
        guard let last = last else { return nil }
        guard Date().timeIntervalSince(last) < ttlSeconds else {
            evict(sessionId)
            return nil
        }
        guard let data = loadFromKeychain(key: keychainKey(sessionId)),
              let obj = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            return nil
        }
        touch(sessionId)
        return obj
    }

    /** Refresh the last-used timestamp, resetting the TTL countdown. */
    func touch(_ sessionId: String) {
        lock.lock()
        if lastUsed[sessionId] != nil {
            lastUsed[sessionId] = Date()
        }
        lock.unlock()
    }

    /** Remove sessionId from the vault and index immediately. */
    func evict(_ sessionId: String) {
        lock.lock()
        lastUsed.removeValue(forKey: sessionId)
        lock.unlock()
        deleteFromKeychain(key: keychainKey(sessionId))
    }

    /** Remove all sessions whose last-used time exceeds the TTL. */
    func evictExpired() {
        lock.lock()
        let expired = lastUsed.filter { Date().timeIntervalSince($0.value) >= ttlSeconds }.map { $0.key }
        lock.unlock()
        expired.forEach { evict($0) }
    }

    /** Remove all sessions from the vault and index. */
    func evictAll() {
        lock.lock()
        let all = Array(lastUsed.keys)
        lock.unlock()
        all.forEach { evict($0) }
    }

    // MARK: - Private

    private func keychainKey(_ sessionId: String) -> String {
        return "\(namespace).\(sessionId)"
    }

    private func baseQuery(key: String) -> [String: Any] {
        return [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: keychainService,
            kSecAttrAccount as String: key,
        ]
    }

    private func saveToKeychain(key: String, data: Data) {
        var query = baseQuery(key: key)
        SecItemDelete(query as CFDictionary)
        query[kSecValueData as String] = data
        query[kSecAttrAccessible as String] = kSecAttrAccessibleWhenUnlocked
        SecItemAdd(query as CFDictionary, nil)
    }

    private func loadFromKeychain(key: String) -> Data? {
        var query = baseQuery(key: key)
        query[kSecReturnData as String] = true
        query[kSecMatchLimit as String] = kSecMatchLimitOne
        var result: AnyObject?
        let status = SecItemCopyMatching(query as CFDictionary, &result)
        return status == errSecSuccess ? result as? Data : nil
    }

    private func deleteFromKeychain(key: String) {
        SecItemDelete(baseQuery(key: key) as CFDictionary)
    }
}

File: android/src/main/java/com/t6x/plugins/networktools/SessionRegistry.kt
package com.t6x.plugins.networktools

import android.content.Context
import android.util.Log
import androidx.security.crypto.EncryptedSharedPreferences
import androidx.security.crypto.MasterKey
import org.json.JSONObject
import java.util.concurrent.ConcurrentHashMap

/**
 * Manages stateful session lifecycles with vaulted credential storage and TTL-based eviction.
 *
 * Credentials are stored in EncryptedSharedPreferences (AndroidKeyStore-backed AES-256-GCM).
 * An in-memory index tracks last-used timestamps for TTL enforcement without storing
 * sensitive data in plain memory structures.
 *
 * Generic by design — reusable for SSH, TCP, WebSocket, or any stateful connection.
 */
class SessionRegistry(
    context: Context,
    private val namespace: String,
    private val ttlMs: Long = 15 * 60 * 1000L
) {
    private val tag = "SessionRegistry[$namespace]"
    private val lastUsed = ConcurrentHashMap<String, Long>()

    private val prefs = EncryptedSharedPreferences.create(
        context,
        "session_registry_$namespace",
        MasterKey.Builder(context)
            .setKeyScheme(MasterKey.KeyScheme.AES256_GCM)
            .build(),
        EncryptedSharedPreferences.PrefKeyEncryptionScheme.AES256_SIV,
        EncryptedSharedPreferences.PrefValueEncryptionScheme.AES256_GCM
    )

    /** Vault the connection params for [sessionId] and record its last-used time. */
    fun register(sessionId: String, params: Map<String, Any?>) {
        evictExpired()
        val filtered = params.filterValues { it != null }
        val json = JSONObject(filtered as Map<*, *>).toString()
        prefs.edit().putString("session_$sessionId", json).apply()
        lastUsed[sessionId] = System.currentTimeMillis()
        Log.d(tag, "Registered session $sessionId")
    }

    /**
     * Retrieve vaulted params for [sessionId].
     * Returns null if the session is unknown or has exceeded the TTL.
     */
    fun getParams(sessionId: String): Map<String, Any?>? {
        evictExpired()
        val last = lastUsed[sessionId] ?: return null
        if (System.currentTimeMillis() - last > ttlMs) {
            Log.d(tag, "Session $sessionId TTL expired")
            evict(sessionId)
            return null
        }
        val json = prefs.getString("session_$sessionId", null) ?: return null
        return try {
            val obj = JSONObject(json)
            obj.keys().asSequence().associateWith { key ->
                val v = obj.get(key)
                if (v == JSONObject.NULL) null else v
            }
        } catch (e: Exception) {
            Log.w(tag, "Failed to parse params for $sessionId: ${e.message}")
            null
        }
    }

    /** Refresh the last-used timestamp, resetting the TTL countdown. */
    fun touch(sessionId: String) {
        if (lastUsed.containsKey(sessionId)) {
            lastUsed[sessionId] = System.currentTimeMillis()
        }
    }

    /** Remove [sessionId] from the vault and index immediately. */
    fun evict(sessionId: String) {
        lastUsed.remove(sessionId)
        prefs.edit().remove("session_$sessionId").apply()
        Log.d(tag, "Evicted session $sessionId")
    }

    /** Remove all sessions whose last-used time exceeds the TTL. */
    fun evictExpired() {
        val now = System.currentTimeMillis()
        val expired = lastUsed.entries
            .filter { now - it.value > ttlMs }
            .map { it.key }
        expired.forEach { id ->
            lastUsed.remove(id)
            prefs.edit().remove("session_$id").apply()
            Log.d(tag, "TTL-evicted session $id")
        }
    }

    /** Remove all sessions from the vault and index. */
    fun evictAll() {
        val all = lastUsed.keys.toList()
        all.forEach { evict(it) }
    }
}
